### ISSUE TITLE
Move KEv2 operators from system-charts to feature-charts

### ISSUE THREAD

**Issue by cbron** [rancher/rancher#31592]
With the migration from v1 to v2 for app deployments, we need to move the these operators from system-charts to feature-charts. 

We need to support the upgrade from system-charts to feature-charts

**Comment by aaronRancher** [comment]
On Rancher `master-head` `version bfdeb22e9`, I am unable to deploy an AKS cluster.

From Ember or Dashboard UI, navigate to AKS cluster create.
Enter cluster resource group and DNS Prefix.
Leave remaining fields default
Create cluster.

Rancher server panics when user tries to deploy an AKS cluster:
```
k8s.io/apimachinery/pkg/util/wait.Until(0xc0032e9420, 0x3b9aca00, 0xc0017016e0)
	/go/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/wait/wait.go:90 +0x4d
created by github.com/rancher/lasso/pkg/controller.(*controller).run
	/go/pkg/mod/github.com/rancher/lasso@v0.0.0-20210616224652-fc3ebd901c08/pkg/controller/controller.go:134 +0x33b
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x34c9f1b]

goroutine 4280 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.21.0/pkg/util/runtime/runtime.go:55 +0x109
panic(0x4899ac0, 0x83202f0)
	/usr/local/go/src/runtime/panic.go:965 +0x1b9
```

**Comment by aaronRancher** [comment]
On Rancher `master-head` `version 6757965`, I see Rancher panic when creating a GKE cluster.

From Ember or Dashboard UI, navigate to GKE cluster create.
Provide cloud credential and node pool name.
Leave all other fields default.

Rancher server panics with the same error as above:
```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0x34ca1bb]
```
Full panic logs: [gke_rancher_panic.txt](https://github.com/rancher/rancher/files/6737115/gke_rancher_panic.txt)

**Comment by thehejik** [comment]
Rancher pod is crashing exactly the same way also when creating `Imported GKE` cluster on `rancher/rancher:master-head c1151f6`.

**Comment by aaronRancher** [comment]
Tested on Rancher `master-head` `version 5b0da3d`

I am able to launch GKEv2, EKSv2, and AKSv2 clusters. System workloads are healthy.
GKE operator version: `eks-operator:v1.1.1-rc5`
EKS operator version: `gke-operator:v1.1.1-rc6`
AKS operator version: `aks-operator:v1.0.1-rc14`

QA to do:
Test upgrade from system-charts to feature-charts

**Comment by rmweir** [comment]
Reopened because deleting the old eks-operator app causes a bad state.

Steps:
* Install rancher v2.5
* Create eks cluster and wait for it to become active.
* Upgrade to rancher v2.6.
* Delete old eks-operator app in old UI in system project.

Results:
* New and old eks-operator apps are deleted and there are no eks-operator deployments.
* Eks-operator CRD app is not deleted.
* EKS cluster edits no longer work.
* EKS cluster edits do not cause app to redeploy.
* CRD app stays the same.

Expected:
Editing an EKS cluster should drive desired updates or produces errors.
PR is up that will just delete the old app and recreate it. I believe there will still be an associated app.project.cattle.io, but deleting it should not cause any issues.

**Comment by sowmyav27** [comment]
To test here is upgrade path from 2.5 (system-charts) to 2.6 (feature-charts)

**Comment by aaronRancher** [comment]
Reproduced the behavior @rmweir mentioned with the following steps:
- Install rancher v2.5
- Create eks cluster and wait for it to become active. (eks-operator v1.0.8)
- Upgrade to rancher v2.6.
- Delete old eks-operator app in old UI in system project.
- Edit the eks cluster (change the logging)

Result:
Cluster does not update.
eks-operator is not redeployed

**Comment by rmweir** [comment]
A case that should be included in testing:
After the migration, when a new version of the chart with a new CRD comes out, the crd app and underlying CRD are properly updated.

Can test this by making a fork of rancher/charts, adding a field to CRD, and pointing the catalog to it in dashboard. This should be done after initial upgrade of a rancher setup that has eks-operator already deployed and an EKS cluster provisioned.

**Comment by aaronRancher** [comment]
### Tested an upgrade from `v2.5.8` to `master-head` `version 94bca99` with the following steps:
- Bring up Rancher `v2.5.8`
- Deploy a GKEv2, EKSv2, and AKSv1 cluster. (`eks-operator:v1.0.8`, `gke-operator:v1.0.1`)
- Verify clusters are available and system workloads are healthy.
- Upgrade the setup to `master-head` `version 94bca99`
- EKS and GKE operators have been upgraded (`eks-operator:v1.1.1-rc8`, `gke-operator:v1.1.1-rc9`)
- Verify clusters can be edited and upgraded.
- Deploy an AKSv2 cluster.
- AKS operator is deployed. (`aks-operator:v1.0.1-rc17`)

Result:
Clusters can be edited and upgraded after an upgrade from v2.5 to v2.6

### Tested deleting eks operator with the following steps:
- Install rancher v2.5
- Create eks cluster and wait for it to become active. (eks-operator v1.0.8)
- Upgrade to rancher v2.6.
- The old eks-operator app is deleted and eks-operator v1.1.1-rc8 is available.
- Delete eks-operator
- Edit the eks cluster (scale the node pool)

Result:
EKS operator is not redeployed. Cluster does not go into updating state.
Provisioning another EKSv2 cluster does not redeploy eks-operator.
Logged an issue to track this: https://github.com/rancher/rancher/issues/33842

**Comment by aaronRancher** [comment]
### Tested CRD update with the following steps:
- Bring up Rancher v2.5.8
- Create an EKS cluster and wait for it to become active. (eks-operator v1.0.8)
- Upgrade to rancher v2.6 (`master-head` `version 94bca99`). The operator is upgraded to `eks-operator:v1.1.1-rc8`
- Navigate to Dashboard > Local > Apps & Marketplace > Chart Repositories. Change the `rancher-charts` chart catalog. Pointed to `http://github.com/rmweir/charts` using `test-crd-upgrade` branch.
- Initiate a chart upgrade by editing the EKS cluster (scale the node pool)

Result:
The eks-operator app and CRD did not upgrade. `rancher-eks-operator-crd:100.0.0+up1.1.1-rc8`
An upgrade is available manually in Dashboard
![Screen Shot 2021-07-30 at 12 00 14 PM](https://user-images.githubusercontent.com/57409052/127718877-25d284af-2928-4d4b-bce1-09b17e2412af.png)

**Comment by rmweir** [comment]
In regards to @aaronRancher findings, it is a requirement that when a new compatible version is available that any kev2 related input trigger its respective operator app to update.

**Comment by deniseschannon** [comment]
Available to test in v2.6.0-rc1

**Comment by StrongMonkey** [comment]
@aaronRancher This is avaiable to test. The behavior has slightly changed so make sure you follow this process https://github.com/rancher/rancher/pull/33879#issuecomment-892240088.

**Comment by aaronRancher** [comment]
### Tested CRD update with the following steps:
On Rancher `v2.6-head` `version 7b9a232`
- Bring up Rancher v2.5.8
- Create an EKS cluster and wait for it to become active. (eks-operator v1.0.8)
- Upgrade to rancher v2.6 (`v2.6-head` `version 7b9a232`). The operator is upgraded to `eks-operator:v1.1.1-rc8`
- Navigate to Dashboard > Local > Apps & Marketplace > Chart Repositories. Change the `rancher-charts` chart catalog. Pointed to `http://github.com/rmweir/charts` using `test-crd-upgrade` branch. Verified a new eks operator version was available.
![Screen Shot 2021-08-06 at 1 09 28 PM](https://user-images.githubusercontent.com/57409052/128565975-d7613f6b-decd-4368-99b3-51fcd0e62abd.png)
- Wait for EKS operator to upgrade to `eks-operator:v1.1.1-rc9`

Result:
EKS operator did not upgrade to `eks-operator:v1.1.1-rc9`. Waited for ~30 min.

Was unable to find the new setting `system-feature-chart-refresh-minutes`
```
> kubectl get setting system-feature-chart-refresh-minutes -o yaml
Error from server (NotFound): settings.management.cattle.io "system-feature-chart-refresh-minutes" not found
```
Was able to find the new setting `system-feature-chart-refresh-seconds` in local cluster:
```
> kubectl get setting system-feature-chart-refresh-seconds -o yaml
apiVersion: management.cattle.io/v3
customized: false
default: "900"
kind: Setting
metadata:
  creationTimestamp: "2021-08-06T19:28:06Z"
  generation: 1
  managedFields:
  - apiVersion: management.cattle.io/v3
    fieldsType: FieldsV1
    fieldsV1:
      f:customized: {}
      f:default: {}
      f:source: {}
      f:value: {}
    manager: rancher
    operation: Update
    time: "2021-08-06T19:28:06Z"
  name: system-feature-chart-refresh-seconds
  resourceVersion: "27044"
  uid: c9071936-edf2-4321-9410-aa72d9b355d4
source: ""
value: ""
```

### Tested deleting eks operator with the following steps:
Using the same setup as above. On Rancher `v2.6-head` `version 7b9a232`
- Delete the EKS operator app and CRD
- Edit an existing EKSv2 cluster. (Scale the node pool)

Result:
- EKS operator is redeployed using latest version. `eks-operator:v1.1.1-rc9` (As chart was still modified in above setup)
![Screen Shot 2021-08-06 at 1 19 03 PM](https://user-images.githubusercontent.com/57409052/128566831-d202c250-1267-40eb-afc0-0f2f39c67096.png)

Reopening as I do not see EKS operator automatically update with interval set in `system-feature-chart-refresh-seconds`

**Comment by aaronRancher** [comment]
Retested CRD update with new chart branch.
Tested with chart `https://github.com/StrongMonkey/charts-1` on `test-crd-upgrade` branch.
The helm version needs to be bumped for the server to recognize a new version.

Steps:
- Bring up Rancher v2.5.8
- Create an EKS cluster and wait for it to become active. (eks-operator v1.0.8)
- Upgrade to rancher v2.6 (v2.6-head version 7b9a232). The operator is upgraded to `rancher-eks-operator:100.0.0+up1.1.1-rc8`
- Change chart catalog
- Wait for EKS operator upgrade

Result
EKS operator CRD is upgraded after ~15 min. Upgrades to `rancher-eks-operator-crd:100.0.1+up1.1.1-rc8`. EKS operator app is not upgraded.
![Screen Shot 2021-08-06 at 3 17 02 PM](https://user-images.githubusercontent.com/57409052/128576227-4542dc67-c285-421c-91ec-d9a21744a8d2.png)

**Comment by aaronRancher** [comment]
Tested GKE and AKS operators on Rancher `master-head` `version 84770f0` and `v2.6-head` `version c47122a`.
GKE and AKS upgrade from system-charts to feature-charts were done in above [validations](https://github.com/rancher/rancher/issues/31592#issuecomment-889552528). Tested with the following steps:
- Bring up a Rancher v2.5.8 setup.
- Deploy a GKEv2 (k8s version 1.20) cluster.
- GKE operator `v1.0.1` is deployed.
- Upgrade Rancher to `v2.6-head` `version c47122a`. Also tested with `master-head` `version 84770f0`.
- GKE operator is automatically updated to `v1.1.1-rc10`
- Deploy an AKSv2 cluster (k8s version 1.20). As AKSv2 is not available in v2.5.8, this cluster is deployed after an upgrade.
- AKS operator `v1.0.1-rc18` is deployed.
- Edit the GKEv2 and AKSv2 clusters (scale the node pool).

Result:
- GKEv2 and AKSv2 clusters can be edited. Cluster edit and deploy functionality still works.
- An AKSv1 cluster deployed on v2.5.8 can still be edited on Rancher upgrade to `v2.6-head` and `master-head`

### Tested deleting GKE and AKS operator with the following steps:
- Using the above setup, delete EKS and AKS operators and CRDs.
- Edit the existing GKEv2 and AKSv2 clusters.

Result:
- EKS operator `v1.1.1-rc10` and AKS operator `v1.0.1-rc18` are redeployed.
- Edit completes and is seen in Rancher UI and backend.

### CRD update with new chart branch tested with the following steps:
- Tested with chart `https://github.com/StrongMonkey/charts-1` on `test-crd-upgrade branch`.
- Using the above setup, edit the chart catalog.
- Wait for GKE and AKS operator upgrade

Result
- AKS and GKE operator app and CRD are updated automatically within ~15 min.
